### ISSUE TITLE
Different number of reads for same pod5 with different models

### ISSUE THREAD

**Issue by KunFang93** [nanoporetech/bonito#409]
Hi,

I tried iterative strategy to finer tuning the model. I trained three models with 
**round0**
```
#!/bin/bash

MODEL_PATH="dna_r10.4.1_e8.2_400bps_hac@v5.0.0"
REFERENCE="/mnt/data/kfang/reference/sacCer3_mm/sacCer3.mmi"
MIN_ACC=0.95

# Create necessary directories if they do not exist
#mkdir -p round1/test

# Loop over parts from 00 to 12
#for i in $(seq -w 0 12)
#do
#  # Create the directory for each part if it doesn't exist
#  mkdir -p "round1/part${i}"
#  bonito basecaller "$MODEL_PATH" \
#    "data_pod5_lig" \
#    --reference "$REFERENCE" \
#    --recursive \
#    --save-ctc \
#    --min-accuracy-save-ctc "$MIN_ACC" \
#    --read-ids "all_part${i}" \
#    > "round0/part${i}/basecalls_${i}.bam"
#done

# Create a comma-separated list of part directories for combining
#combine_list=$(seq -w 0 12 | sed 's/^/round1\/part/' | paste -sd "," -)
#python bonito_basecall_parts_combine_large.py -s "$combine_list" -o ./round0_combine

bonito train --epochs 1 --lr 5e-4 --directory ./round0_combine/ ./round0_combine/fine-tuned-model --pretrain "$MODEL_PATH"
```

**round1**
```
#!/bin/bash

MODEL_PATH="./round0_combine/fine-tuned-model/"
REFERENCE="/mnt/data/kfang/reference/sacCer3_mm/sacCer3.mmi"
MIN_ACC=0.95

# Create necessary directories if they do not exist
#mkdir -p round1/test

# Loop over parts from 00 to 12
#for i in $(seq -w 0 12)
#do
#  # Create the directory for each part if it doesn't exist
#  mkdir -p "round1/part${i}"
#  bonito basecaller "$MODEL_PATH" \
#    "data_pod5_lig" \
#    --reference "$REFERENCE" \
#    --recursive \
#    --save-ctc \
#    --min-accuracy-save-ctc "$MIN_ACC" \
#    --read-ids "all_part${i}" \
#    > "round1/part${i}/basecalls_${i}.bam"
#done

# Create a comma-separated list of part directories for combining
#combine_list=$(seq -w 0 12 | sed 's/^/round1\/part/' | paste -sd "," -)
#python bonito_basecall_parts_combine_large.py -s "$combine_list" -o ./round1_combine

bonito train --epochs 1 --lr 5e-4 --directory ./round1_combine/ ./round1_combine/fine-tuned-model --pretrain "$MODEL_PATH"
```

And I exported round0_combine/fine-tuned-model and round1_combine/fine-tune-model with bonito export
```
bonito export --output dna_bonito_model_r0 round0_combine/fine-tuned-model/
bonito export --output dna_bonito_model_r1 round1_combine/fine-tuned-model/
```

When I check the performance of different models, I found that the total number of reads are different from different model:
**dorado official support sup model** 2297227 in total
```
dorado basecaller sup,5mC_5hmC,6mA --kit-name SQK-RBK114-96 -r --output-dir ./ --reference sacCer3.mmi --mm2-opts "-k 15 -w 10" pod5/

samtools flagstat -@ 40 ../calls_2025-02-22_T21-36-58.bam
2297227 + 0 in total (QC-passed reads + QC-failed reads)
1894467 + 0 primary
327805 + 0 secondary
74955 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
1717024 + 0 mapped (74.74% : N/A)
1314264 + 0 primary mapped (69.37% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)
```
**round0_model** 1880743 in total
```
dorado basecaller /dna_bonito_model_r0 --kit-name SQK-RBK114-96 -r --output-dir ./ --reference sacCer3.mmi --mm2-opts "-k 15 -w 10" pod5/

samtools flagstat -@ 40 calls_2025-03-12_T15-08-54.bam
1880743 + 0 in total (QC-passed reads + QC-failed reads)
1554593 + 0 primary
266328 + 0 secondary
59822 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
1589597 + 0 mapped (84.52% : N/A)
1263447 + 0 primary mapped (81.27% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)
```
and **round1_model** 1753039 in total
```
dorado basecaller /dna_bonito_model_r1 --kit-name SQK-RBK114-96 -r --output-dir ./ --reference sacCer3.mmi --mm2-opts "-k 15 -w 10" pod5/

samtools flagstat -@ 40/bonito_r1/calls_2025-03-12_T15-45-24.bam
1753039 + 0 in total (QC-passed reads + QC-failed reads)
1420365 + 0 primary
265031 + 0 secondary
67643 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
1595665 + 0 mapped (91.02% : N/A)
1262991 + 0 primary mapped (88.92% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)
```

I wondered if 
1. these differences are as expected? 
2. what's the potential cause for these?
3. how could I mitigate the reduction of the total reads number?
4. the procedure of training looks reasonable?

Apologize for so many questions. Thank you so much for helping!

Best,
Kun

**Comment by davidnewman02** [comment]
What is your aim in finetuning a model using bonito? 
In the vast majority of cases we expect that the models provided in dorado will basecall a wide variety of data to a high degree of accuracy without any further finetuning, which can be a difficult process to get right.

In this case it looks like the model you have trained is aligning fewer reads than the production hac-v5 model. What do you observe in the alignment accuracy?

**Comment by KunFang93** [comment]
Thanks for your reply! We have two artificial modifications and have concerns about accuracy for regions enriched with both modifications. We did some extreme experiments and sequenced. We confirmed that the alignment accuracy went down if the two modifications are very enriched (ecdf plot attached below) and multiple round of training improve the accuracy.

<img width="252" alt="Image" src="https://github.com/user-attachments/assets/bdde3828-04da-4d0e-bcc4-fcecfa10fa4d" />

As the attached plot, the accuracy is increased but we did see less read number after bonito training. I contacted with dorado team and they thought it was because the build-in split and remove short reads functions. I investigated those missed reads and found most of those reads are **unaligned** reads under dorado sup model. Please let me know if you would like to see the detailed analysis so I could sent to you email for your further investigation.

**Comment by KunFang93** [comment]
Thanks for your suggestions! For my application, accuracy is the most important consideration. I had looked the sequence length and alignment length as below. It looks like those missed reads have certain length but indeed relatively short. 

<img width="886" alt="Image" src="https://github.com/user-attachments/assets/97ef0e81-d7c2-4586-8c17-21de624755a8" />

I’ll take a closer look at potential need for padding in the reference.
